### PR TITLE
shiritori: switch interact to promise

### LINKED ISSUE #9: switch shiritori.interact to promise
To better fit aoglib v2 dispatch mecanism.

### PR DESCRIPTION
**Pull request by proppy**
Fixes #9

### CODE REVIEW

File: functions/shiritori/main.js [hunk 1]
```diff
@@ -27,30 +27,30 @@ const rl = readline.createInterface({
...
+    shiritori
+        .interact(
+            kana => Promise.resolve(corpus[kana]),
+            input,
+            chain,
+        )
+        .then(result => {
+          console.log(`${result.word} [${result.kana}]`)
+          chain.unshift(input)
+          chain.unshift(result.word)
```
Comment (proppy) [inline functions/shiritori/main.js]: this should be `result.kana`

File: functions/shiritori/index.js [hunk 2]
```diff
@@ -98,29 +98,29 @@ exports.kanas = (word) => new Set((() => {
 })())
 
 // しりとりのゲームループ。
-exports.interact = (dict, word, previousInputs, callbacks) => {
-  const next = exports.kanas(word)
-  if ((next.size === 0) || !exports.check(word, previousInputs)) {
-    return callbacks.lose()
+exports.interact = (dict, word, chain) => new Promise((resolve, reject) => {
+  const next = exports.kanas(word);
```
Comment (yoichiro) [inline functions/shiritori/index.js]: Could you remove unnecessary `;` character?

File: functions/shiritori/index.js [hunk 3]
```diff
@@ -98,29 +98,29 @@ exports.kanas = (word) => new Set((() => {
...
+      return reject({win: true})
     }
     const w = unused[Math.floor(Math.random() * unused.length)]
     const wk = words[w]
     if (wk[wk.length - 1] === 'ん') {
-      return callbacks.win(w, wk)
+      return reject({win: true, 'word': w, 'kana': wk})
     }
-    return callbacks.next(w, wk)
+    return resolve({'word': w, 'kana': wk})
```
Comment (yoichiro) [inline functions/shiritori/index.js]: I think that this `return` is unnecessary, therefore we can remove this. How about you?

File: functions/shiritori/index.js [hunk 4]
```diff
@@ -98,29 +98,29 @@ exports.kanas = (word) => new Set((() => {
...
     }
     if (unused.length === 0) {
-      return callbacks.win()
+      return reject({win: true})
     }
     const w = unused[Math.floor(Math.random() * unused.length)]
     const wk = words[w]
     if (wk[wk.length - 1] === 'ん') {
-      return callbacks.win(w, wk)
+      return reject({win: true, 'word': w, 'kana': wk})
```
Comment (yoichiro) [inline functions/shiritori/index.js]: For a unified syntax, I think that this `'word': w, 'kana': wk` should be `word: w, kana: wk`. What reason is there to use key name as string?

File: functions/shiritori/index.js (hunk 3 above)
Comment (yoichiro) [reply functions/shiritori/index.js]: I have the same opinion (key names as string) for this line.

File: functions/shiritori/index.js [hunk 5]
```diff
@@ -98,29 +98,29 @@ exports.kanas = (word) => new Set((() => {
...
         }
       }
     }
     if (unused.length === 0) {
-      return callbacks.win()
+      return reject({win: true})
     }
     const w = unused[Math.floor(Math.random() * unused.length)]
     const wk = words[w]
     if (wk[wk.length - 1] === 'ん') {
```
Comment (yoichiro) [inline functions/shiritori/index.js]: I would like to confirm to be safe. Is the length of the `wk` value guaranteed to 1 or more?

Comment (proppy) [reply functions/shiritori/index.js]: Done

File: functions/shiritori/index.js (hunk 4 above)
Comment (proppy) [reply functions/shiritori/index.js]: Done

File: functions/shiritori/index.js (hunk 3 above)
Comment (proppy) [reply functions/shiritori/index.js]: Done

File: functions/shiritori/index.js (hunk 2 above)
Comment (proppy) [reply functions/shiritori/index.js]: Done

File: functions/shiritori/main.js (hunk 1 above)
Comment (proppy) [reply functions/shiritori/main.js]: Done

**Comment by proppy** [comment]
PTAL

File: functions/index.js [hunk 6]
```diff
@@ -49,33 +49,31 @@ app.intent('actions.intent.MAIN', (conv) => {
   conv.ask('どうぞ、始めて下さい');
 });
 
-app.intent('actions.intent.TEXT', (conv, input) => new Promise((resolve, reject) => {
+app.intent('actions.intent.TEXT', (conv, input) => {
```
[the remark this answers is not in the document]
Comment (proppy) [reply functions/index.js]: Yes, I forgot to commit and push this change: 0c86fbc

Nice catch!

[the remark this answers is not in the document]
Comment (proppy) [reply functions/index.js]: Note that it was already throwing this error, because the db looking is async :)

File: functions/index.js [hunk 7]
```diff
@@ -49,33 +49,31 @@ app.intent('actions.intent.MAIN', (conv) => {
...
-        conv.ask(new SimpleResponse({
-          speech: word,
-          text: `${word} [${kana}]`
-        }));
-        resolve();
-      }
-    });
+app.intent('actions.intent.TEXT', (conv, input) => {
+  return shiritori.loaded.then(() => {
+    return shiritori.interact(dict, input, conv.data.used)
```
Comment (yoichiro) [inline functions/index.js]: 👍

**Pull request merged**